service/history/task/cross_cluster_task_processor_test.go

```diff
@@ -0,0 +1,292 @@
+	fetcher := newTaskFetcher(
+		cluster.TestAlternativeClusterName,
+		cluster.TestCurrentClusterName,
+		nil, nil, nil, nil,
+	)
+	processor := newCrossClusterTaskProcessor(s.mockShard, s.mockProcessor, fetcher, s.processorOptions)
+	numPendingTasks := 10
+	completedTasks := 6
+	futureSettables := make([]future.Settable, numPendingTasks)
+	for i := 0; i != numPendingTasks; i++ {
+		taskFuture, settable := future.NewFuture()
+		processor.pendingTasks[int64(i)] = taskFuture
+		futureSettables[i] = settable
+	}
+
+	for idx := range rand.Perm(numPendingTasks)[:completedTasks] {
+		futureSettables[idx].Set(types.CrossClusterTaskResponse{TaskID: int64(idx)}, nil)
+	}
+
+	s.mockShard.Resource.HistoryClient.EXPECT().RespondCrossClusterTasksCompleted(gomock.Any(), gomock.Any()).DoAndReturn(
+		func(_ context.Context, request *types.RespondCrossClusterTasksCompletedRequest) (*types.RespondCrossClusterTasksCompletedResponse, error) {
+			select {
+			case <-processor.shutdownCh:
+			default:
+				close(processor.shutdownCh)
+			}
+
+			s.Len(request.TaskResponses, completedTasks)
+			s.Equal(s.mockShard.Resource.GetClusterMetadata().GetCurrentClusterName(), request.TargetCluster)
+			s.Equal(s.mockShard.GetShardID(), int(request.GetShardID()))
+			// s.False(request.GetFetchNewTasks())
+			if failedToRespond {
+				return nil, errors.New("some random error")
+			}
+			return &types.RespondCrossClusterTasksCompletedResponse{}, nil
+		},
+	).AnyTimes()
+
+	processor.shutdownWG.Add(1)
+	go processor.respondPendingTaskLoop()
+	processor.shutdownWG.Wait()
+
+	if failedToRespond {
+		s.Len(processor.pendingTasks, numPendingTasks)
+	} else {
+		s.Len(processor.pendingTasks, numPendingTasks-completedTasks)
+	}
+}
+
+func (s *crossClusterTaskProcessorSuite) TestProcessTaskRequests_RespondFailed() {
+	s.testProcessTaskRequests(true)
+}
+
+func (s *crossClusterTaskProcessorSuite) TestProcessTaskRequests_RespondSuccess() {
+	s.testProcessTaskRequests(false)
+}
+
+func (s *crossClusterTaskProcessorSuite) testProcessTaskRequests(failedToRespond bool) {
+	fetcher := newTaskFetcher(
+		cluster.TestAlternativeClusterName,
+		cluster.TestCurrentClusterName,
+		nil, nil, nil, nil,
+	)
+	processor := newCrossClusterTaskProcessor(s.mockShard, s.mockProcessor, fetcher, s.processorOptions)
+	pendingTaskID := 0
+	future, _ := future.NewFuture()
+	processor.pendingTasks[int64(pendingTaskID)] = future
+
+	numTasks := 10
+	var tasksRequests []*types.CrossClusterTaskRequest
+	for id := pendingTaskID; id != pendingTaskID+numTasks; id++ {
+		tasksRequests = append(tasksRequests, &types.CrossClusterTaskRequest{
+			TaskInfo: &types.CrossClusterTaskInfo{
+				TaskID: int64(id),
+			},
+		})
+	}
+
+	completedTasks := 0
+	s.mockProcessor.EXPECT().TrySubmit(gomock.Any()).DoAndReturn(
+		func(t Task) (bool, error) {
+			submitted := rand.Intn(2) == 0
+			if submitted {
+				completedTasks++
+				t.Ack()
+			}
+			// redispatcher interval is set to 1hr, basically disabled
```

Comment:
If a task is not accepted by the mockProcessor, it will be submitted to the redispatcher so that the submission can be retried later. 

For this test I set the time interval for redispatch to 1hr which basically disabled redispatcher so that I can easily count how many tasks are submitted/completed by the mockProcessor. 
If the redispatcher is working and re-submitting tasks from time to time, then the logic for counting exactly how many tasks are submitted will be more complicated.